src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -261,15 +262,25 @@ public void write() {
               }
             } else {
               SinkRecord projectedRecord = SchemaUtils.project(record, currentSchema, compatibility);
-              writeRecord(projectedRecord);
-              buffer.poll();
-              if (shouldRotate(now)) {
+              String partition = partitioner.encodePartition(projectedRecord);
+              if (currentPartition != null && !currentPartition.equals(partition)) {
+                currentPartition = partition;
```

Comment:
Does the `rotate.interval.ms` setting not work for you? The two conditions currently checked in `shouldRotate` are whether there are enough messages or if enough time has elapsed.